### PR TITLE
add support for Eventer along with all of its dependencies

### PR DESCRIPTION
**Pull request by jimlambrt**
This is second of 3 PRs to add the events package and they should be reviewed in order.  Before this was #1275 and after this one there's #1277.

This PR will be merged into the branch for #1275, which will be merged to main when all three have been reviewed/approved.

### CODE REVIEW

File: internal/observability/events/eventer_retry.go [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+}
+
+type retryInfo struct {
+	retries int
+	backoff time.Duration
+}
+
+type sendHandler func() (eventlogger.Status, error)
+
+func (e *Eventer) retrySend(ctx context.Context, retries uint, backOff backoff, handler sendHandler) error {
```
Comment (s-christoff) [inline internal/observability/events/eventer_retry.go:38]: I feel like this would do well with some comments

**Review by s-christoff — COMMENTED** [review]
Looking great!

File: internal/observability/events/eventer.go [hunk 2]
```diff
@@ -9,4 +21,378 @@ const (
...
+		logger: log,
+		conf:   c,
+		broker: b,
+	}
+
+	if !opts.withNow.IsZero() {
+		e.broker.StopTimeAt(opts.withNow)
+	}
+
+	// Create JSONFormatter node
```
[the remark this answers is not in the document]
Comment (jimlambrt) [reply internal/observability/events/eventer.go:134]: Hum... I get your point, and I know you have some personal experience with not adding a formatting node to a pipeline, but I'm not sure there's anything in here that's unnecessary.  For example, if we don't add the error file sink, no errors events will be written.

File: internal/observability/events/eventer.go [hunk 3]
```diff
@@ -9,4 +21,378 @@ const (
...
+	return nil
+}
+
+// SysEventer returns the "system wide" eventer for Boundary.
+func SysEventer() *Eventer {
+	return sysEventer
+}
+
+// NewEventer creates a new Eventer using the config.  Supports options: WithNow
+func NewEventer(log hclog.Logger, c EventerConfig, opt ...Option) (*Eventer, error) {
```
[the remark this answers is not in the document]
Comment (jimlambrt) [reply internal/observability/events/eventer.go:87]: It's possible...  but I don't think it's typically helpful decompose functions into multiple unexported functions which are not reusable. I will consider your suggestion a bit more though.

File: internal/observability/events/eventer.go [hunk 4]
```diff
@@ -9,4 +21,378 @@ const (
...
 )
+
+// flushable defines an interface that all eventlogger Nodes must implement if
+// they are "flushable"
+type flushable interface {
+	FlushAll(ctx context.Context) error
+}
+
+// broker defines an interface for an eventlogger Broker... which will allow us
+// to substitute our our testing broker when needed to write tests for things
```
Comment (louisruch) [inline internal/observability/events/eventer.go:37]:
```suggestion
// to substitute our testing broker when needed to write tests for things
```

File: internal/observability/events/eventer.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (louisruch) [reply internal/observability/events/eventer.go:87]: +1 This is a pretty large function

File: internal/observability/events/eventer.go [hunk 5]
```diff
@@ -9,4 +21,378 @@ const (
...
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to set success threshold for error events")
+	}
+
+	return e, nil
+}
+
+// writeObservation writes/sends an Observation event.
+func (e *Eventer) writeObservation(ctx context.Context, event *observation) error {
+	const op = "event.(Eventer).WriteObservation"
```
Comment (louisruch) [inline internal/observability/events/eventer.go:320]:
```suggestion
	const op = "event.(Eventer).writeObservation"
```

File: internal/observability/events/eventer.go [hunk 6]
```diff
@@ -9,4 +21,378 @@ const (
...
+	if err != nil {
+		e.logger.Error("encountered an error sending an observation event", "error:", err.Error())
+		return errors.Wrap(err, op)
+	}
+	return nil
+}
+
+// writeError writes/sends an Err event
+func (e *Eventer) writeError(ctx context.Context, event *err) error {
+	const op = "event.(Eventer).WriteError"
```
Comment (louisruch) [inline internal/observability/events/eventer.go:346]:
```suggestion
	const op = "event.(Eventer).writeError"
```

File: internal/observability/events/eventer.go [hunk 7]
```diff
@@ -9,4 +21,378 @@ const (
...
+	if err != nil {
+		e.logger.Error("encountered an error sending an error event", "error:", err.Error())
+		return errors.Wrap(err, op)
+	}
+	return nil
+}
+
+// writeAudit writes/send an audit event
+func (e *Eventer) writeAudit(ctx context.Context, event *Audit) error {
+	const op = "event.(Eventer).WriteAudit"
```
Comment (louisruch) [inline internal/observability/events/eventer.go:362]:
```suggestion
	const op = "event.(Eventer).writeAudit"
```

File: internal/observability/events/eventer.go [hunk 8]
```diff
@@ -9,4 +21,378 @@ const (
...
+	if e.broker != nil {
+		return e.broker.Reopen(context.Background())
+	}
+	return nil
+}
+
+// FlushNodes will flush any of the eventer's flushable nodes.  This
+// needs to be called whenever Boundary is stopping (aka shutting down).
+func (e *Eventer) FlushNodes(ctx context.Context) error {
+	const op = "event.(Eventer).FlushAll"
```
Comment (louisruch) [inline internal/observability/events/eventer.go:391]:
```suggestion
	const op = "event.(Eventer).FlushNodes"
```

File: internal/observability/events/eventer_retry.go [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+
+type backoff interface {
+	duration(attemptNumber uint) time.Duration
+}
+
+type expBackoff struct{}
+
+func (b expBackoff) duration(attempt uint) time.Duration {
+	r := rand.Float64()
+	return time.Millisecond * time.Duration(math.Exp2(float64(attempt))*5*(r+0.5))
```
Comment (louisruch) [inline internal/observability/events/eventer_retry.go:28]: Might be nice to have a comment here or on the method

File: internal/observability/events/eventer_retry.go [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+	if handler == nil {
+		return errors.New(errors.InvalidParameter, op, "missing handler")
+	}
+	success := false
+	var retryErrors error
+	var attemptStatus eventlogger.Status
+	info := retryInfo{}
+	for attempts := uint(1); ; attempts++ {
+		if attempts > retries+1 {
+			retryErrors = multierror.Append(retryErrors, errors.New(errors.MaxRetries, op, fmt.Sprintf("Too many retries: %d of %d", attempts-1, retries+1)))
```
Comment (louisruch) [inline internal/observability/events/eventer_retry.go:52]: This is probably fine and me being super nitty, but it seems strange to call `retrySend` with the retries parameter as `3` and then get an error message: `Too many retries: 4 of 4`

File: internal/observability/events/options.go [hunk 11]
```diff
@@ -27,6 +27,10 @@ type options struct {
 	withRequest     *Request
 	withResponse    *Response
 	withAuth        *Auth
+
+	withBroker          broker // test only option
```
Comment (louisruch) [inline internal/observability/events/options.go:31]: minor nit: maybe `withTestBroker`

**Review by louisruch — COMMENTED** [review]
Some comments and suggestion but this is looking great, looking forward to having all of this in Boundary!

File: internal/observability/events/eventer_retry.go (hunk 9 above)
Comment (jimlambrt) [reply internal/observability/events/eventer_retry.go:29]: done.

File: internal/observability/events/options.go (hunk 11 above)
Comment (jimlambrt) [reply internal/observability/events/options.go:31]: umm... it actually is anything that implements the `broker` interface.

File: internal/observability/events/eventer_retry.go (hunk 1 above)
Comment (jimlambrt) [reply internal/observability/events/eventer_retry.go:39]: Any specific comments?  How about:  `Hi Mom!  I hope things are well and I miss you.  Today, Finn and I took a nap after lunch.  We really should do that every day.  BTW, Sarah was working on her day off.  She'd be giving me so much grief if I did that.  Too funny!  Well, gotta go.  Love you Mom!`

Comment (jimlambrt) [reply internal/observability/events/eventer_retry.go:41]: I added a more appropriate comment.  ty!

File: internal/observability/events/eventer_retry.go (hunk 10 above)
Comment (jimlambrt) [reply internal/observability/events/eventer_retry.go:55]: if retries == 3 is it failed after 3 attempts or failed after 4 attempts? does the UX imply that if 0 is passed for retries that no attempt will be made or just 1? This is what I was trying to disambiguate unsuccessfully I guess. At this point, I'm punting on even trying to conveue the actual number of attempts and just saying being as vauge as possible and returning the reached max of %d where %d is the number of retries.

File: internal/observability/events/event_delivery_guarantee.go [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+const (
+	DefaultDeliveryGuarantee DeliveryGuarantee = ""            // DefaultDeliveryGuarantee will be BestEffort
+	Enforced                 DeliveryGuarantee = "enforced"    // Enforced means that a delivery guarantee is enforced
+	BestEffort               DeliveryGuarantee = "best-effort" // BestEffort means that a best effort will be made to deliver an event
+)
+
+type DeliveryGuarantee string // DeliveryGuarantee defines the guarantees around delivery of an event type within config
+
+func (g DeliveryGuarantee) validate() error {
+	const op = "event.(DeliveryGuarantee"
```
[the remark this answers is not in the document]
Comment (jimlambrt) [reply internal/observability/events/event_delivery_guarantee.go:18]: ty!

[the remark this answers is not in the document]
Comment (jimlambrt) [reply internal/observability/events/event_delivery_guarantee.go]: ty!

File: internal/observability/events/eventer.go [hunk 13]
```diff
@@ -9,4 +21,378 @@ const (
...
+		b = eventlogger.NewBroker()
+	}
+
+	e := &Eventer{
+		logger: log,
+		conf:   c,
+		broker: b,
+	}
+
+	if !opts.withNow.IsZero() {
```
[the remark this answers is not in the document]
Comment (jimlambrt) [reply internal/observability/events/eventer.go:130]: we disagree.

File: internal/observability/events/eventer.go [hunk 14]
```diff
@@ -9,4 +21,378 @@ const (
...
+	}
+	sysEventerOnce.Do(func() {
+		sysEventer = eventer
+	})
+	return nil
+}
+
+// SysEventer returns the "system wide" eventer for Boundary.
+func SysEventer() *Eventer {
+	return sysEventer
```
[the remark this answers is not in the document]
Comment (jimlambrt) [reply internal/observability/events/eventer.go:83]: See the event.WriteErro(...) in next PR... yes, there are valid cases where we want to return it uninitialized.

File: internal/observability/events/eventer.go [hunk 15]
```diff
@@ -9,4 +21,378 @@ const (
...
+	}
+
+	e := &Eventer{
+		logger: log,
+		conf:   c,
+		broker: b,
+	}
+
+	if !opts.withNow.IsZero() {
+		e.broker.StopTimeAt(opts.withNow)
```
[the remark this answers is not in the document]
Comment (jimlambrt) [reply internal/observability/events/eventer.go:131]: Still futzing with a use-case that includes coverage in the WIP branch so I'm not ready to remove it.

File: internal/observability/events/eventer_test.go [hunk 16]
```diff
@@ -0,0 +1,492 @@
...
+			require.NoError(err)
+			require.NotNil(got)
+			tt.want.broker = got.broker
+			tt.want.flushableNodes = got.flushableNodes
+			assert.Equal(tt.want, got)
+		})
+	}
+}
+
+func TestEventer_writeObservation(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (jimlambrt) [reply internal/observability/events/eventer_test.go:86]: I have end to end coverage for output into sinks in the final PR which covers this.

File: internal/observability/events/testing.go [hunk 17]
```diff
@@ -1,13 +1,123 @@
...
+	ObservationEvents *os.File
+	AuditEvents       *os.File
+}
+
+// TestEventerConfig creates a test config and registers a cleanup func for its
+// test tmp files.
+func TestEventerConfig(t *testing.T, testName string, opt ...Option) TestConfig {
+	t.Helper()
+	require := require.New(t)
+	tmpAllFile, err := ioutil.TempFile("./", "tmp-all-events-"+testName)
```
[the remark this answers is not in the document]
Comment (jimlambrt) [reply internal/observability/events/testing.go:42]: See my previous comment: I have end to end coverage for output into sinks in the final PR which covers this.

File: internal/observability/events/eventer_retry.go [hunk 18]
```diff
@@ -0,0 +1,80 @@
...
+			return retryErrors
+		}
+		var err error
+		attemptStatus, err = handler()
+		if len(attemptStatus.Warnings) > 0 {
+			var retryWarnings error
+			for _, w := range attemptStatus.Warnings {
+				retryWarnings = multierror.Append(retryWarnings, w)
+			}
+			e.logger.Error("unable to send event", "operation", op, "warning", retryWarnings)
```
[the remark this answers is not in the document]
Comment (jimlambrt) [reply internal/observability/events/eventer_retry.go:65]: I'm not testing that hclog works.  I think, that's probably sufficiently tested at this point.  The end to end tests in the final PR include validation that the data is written to the sinks.

File: internal/observability/events/eventer_retry.go [hunk 19]
```diff
@@ -0,0 +1,80 @@
...
+			info.retries++
+			info.backoff = info.backoff + d
+			time.Sleep(d)
+			continue
+		}
+		success = true
+		break
+	}
+	if !success {
+		return errors.Wrap(retryErrors, op, errors.WithMsg("failed to send event"))
```
[the remark this answers is not in the document]
Comment (jimlambrt) [reply internal/observability/events/eventer_retry.go:79]: I believe it's an impossible condition as currently written but included for the sake of completeness and as "defense in depth" if someone in the future changes how the loop is implemented.  With that said, I'm not sure as written it's possible to cover this line.

File: internal/observability/events/eventer.go [hunk 20]
```diff
@@ -9,4 +21,378 @@ const (
...
+}
+
+// writeAudit writes/send an audit event
+func (e *Eventer) writeAudit(ctx context.Context, event *Audit) error {
+	const op = "event.(Eventer).writeAudit"
+	if event == nil {
+		return errors.New(errors.InvalidParameter, op, "missing event")
+	}
+	if !e.conf.AuditEnabled {
+		return nil
```
[the remark this answers is not in the document]
Comment (jimlambrt) [reply internal/observability/events/eventer.go:367]: In the final PR where output is validated.

**Review by louisruch — APPROVED** [review]
LGTM! this all looks great 🚢

File: internal/observability/events/eventer.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (vancluever) [reply internal/observability/events/eventer.go:130]: Reached out offline to @jimlambrt  to discuss. The main intent here is more idiomatic and _not_ to specifically take the time at option time. :+1:

File: internal/observability/events/testing.go [hunk 21]
```diff
@@ -53,3 +163,68 @@ func testResponse(t *testing.T) *Response {
 		},
 	}
 }
+
+// testWithBroker is an unexported and a test option for passing in an optional broker
+func testWithBroker(b broker) Option {
```
[the remark this answers is not in the document]
Comment (vancluever) [reply internal/observability/events/testing.go:168]: Just mainly noting that maybe _if_ they are purely test options and will never be non-test options, that maybe they just be named as such. This is minor though for sure, so not a big deal to keep the way it is. :+1:

**Review by vancluever — APPROVED** [review]
LGTM here!

File: internal/observability/events/testing.go (hunk 21 above)
[the remark this answers is not in the document]
Comment (louisruch) [reply internal/observability/events/testing.go:168]: This was the original reason for my request to add `test` to the withBroker, but agree with Chris this is minor and should not block this merging!

**Pull request merged**